### PR TITLE
IMPRO-840 radar coverage masks

### PR DESCRIPTION
**Pull request by cgsandford**
Change to advection plugin to handle masked input data.  This changes the KGO for extrapolation nowcasting to preserve the "out of range" mask that comes in with the radar composite to mask out the Nimrod "no data" fill value value.

Testing:
 - [x] Ran tests and they passed OK
 - [ ] Added new tests for the new feature(s)

### CODE REVIEW

File: lib/improver/nowcasting/optical_flow.py [hunk 1]
```diff
@@ -248,9 +255,6 @@ def process(self, cube, timestep, fill_value=0.0):
                 The 2D cube containing data to be advected
             timestep (datetime.timedelta):
                 Advection time step
-            fill_value (float):
-                Default output value for spatial points where data cannot be
-                extrapolated (source is out of bounds)
 
         Returns:
             advected_cube (iris.cube.Cube):
```
Comment (MoseleyS) [inline lib/improver/nowcasting/optical_flow.py]: I think the doc for the returned "advected_cube" should be updated to specify that points where source was out-of-bounds in the extrapolation will be returned in a masked state. Currently, this info is buried in the doc-string of a hidden method.

File: lib/improver/nowcasting/optical_flow.py [hunk 2]
```diff
@@ -226,16 +224,25 @@ def point_in_bounds(x, y, nx, ny):
         x_weights = [1. - x_weight_upper, x_weight_upper]
         y_weights = [1. - y_weight_upper, y_weight_upper]
 
+        # Check whether the input data is masked - if so substitute NaNs for
+        # the masked data
+        if isinstance(data, np.ma.MaskedArray):
+            data = np.where(data.mask, np.nan, data.data)
```
Comment (bayliffe) [inline lib/improver/nowcasting/optical_flow.py]: This includes an implicit type conversion if the input is an integer; np.nan cannot be stored in a integer array, so it becomes a float array. I don't think this is ever going to be an issue, but thought I would mention it for interest.

File: lib/improver/nowcasting/optical_flow.py [hunk 3]
```diff
@@ -226,16 +224,25 @@ def point_in_bounds(x, y, nx, ny):
...
             for ypt, ywt in zip(y_points, y_weights):
                 cond = point_in_bounds(xpt, ypt, xdim, ydim) & cond_pt
-                self._increment_output_array(data, adv_field, cond, xgrid,
-                                             ygrid, xpt, ypt, xwt, ywt)
+                self._increment_output_array(
+                    data, adv_field, cond, xgrid, ygrid, xpt, ypt, xwt, ywt)
+
+        # Replace NaNs with a mask
+        adv_mask = ~np.isfinite(adv_field)
+        adv_field = np.ma.MaskedArray(adv_field, mask=adv_mask)
```
Comment (bayliffe) [inline lib/improver/nowcasting/optical_flow.py]: Or in one line and without the need for the variable adv_mask:
```
adv_field = np.ma.masked_where(~np.isfinite(adv_field), adv_field)
```

File: lib/improver/tests/nowcasting/optical_flow/test_AdvectField.py [hunk 4]
```diff
@@ -163,41 +164,63 @@ def setUp(self):
...
+        expected_output = np.array([[np.nan, np.nan, np.nan],
+                                    [np.nan, np.nan, np.nan],
+                                    [np.nan, 2., 3.],
+                                    [np.nan, 1., 2.]])
         result = self.dummy_plugin._advect_field(
-            self.data, self.grid_vel_x, self.grid_vel_y, self.timestep, 0.)
-        self.assertArrayAlmostEqual(result, expected_output)
+            self.data, self.grid_vel_x, self.grid_vel_y, self.timestep)
+        self.assertArrayAlmostEqual(result[~result.mask],
+                                    expected_output[~result.mask])
```
Comment (bayliffe) [inline lib/improver/tests/nowcasting/optical_flow/test_AdvectField.py]: Why, in this case and all the subsequent ones, are you only checking that 4/6/some sample of values are correct?

I can see that implicitly it means the masking is correct on the returned cube, so we can interpret that there must be nans at the other points. However, could we make the tests easier to understand by just comparing ```result.data.data``` against the ```expected_output``` array? And then compare the ```result.data.mask``` against ```~np.isfinite(expected_output)```.

That very explicitly checks the data and the mask are correct, requiring less interpretation when someone comes back to look at this.

Comment (cgsandford) [reply lib/improver/tests/nowcasting/optical_flow/test_AdvectField.py]: Ah I had another thought: I could use:
```
self.assertTrue(np.allclose(result.data.data, expected_data, equal_nan=True))
```
A little non-standard but removes the mask / unmask confusion.  @bayliffe what's your preference?

Comment (bayliffe) [reply lib/improver/tests/nowcasting/optical_flow/test_AdvectField.py]: I must have missed the other test. Leave them as is if Stephen is happy.

**Pull request merged**